File: src/api/phrases.ts
import { invoke } from "@tauri-apps/api/core";
import type { PhraseItem } from "./types";

export type AddPhraseInput = {
  phrase: string;
  contextSentence: string;
  articleId?: string | null;
  meaningZh?: string | null;
  usage?: string | null;
};

export const apiPhrases = {
  addPhrase: (input: AddPhraseInput) =>
    invoke<PhraseItem>("add_phrase", {
      input: {
        phrase: input.phrase,
        context_sentence: input.contextSentence,
        article_id: input.articleId ?? null,
        meaning_zh: input.meaningZh ?? null,
        usage: input.usage ?? null,
      },
    }),
  listPhrases: (status?: string) =>
    invoke<PhraseItem[]>("list_phrases", { status: status ?? null }),
  duePhrases: () => invoke<PhraseItem[]>("due_phrases"),
  reviewPhrase: (id: string, rating: string) =>
    invoke<PhraseItem>("review_phrase", { id, rating }),
  setPhraseStatus: (id: string, status: string) =>
    invoke<void>("set_phrase_status", { id, status }),
  deletePhrase: (id: string) => invoke<void>("delete_phrase", { id }),
};

File: src-tauri/src/commands/phrases.rs
use crate::config;
use crate::db::{self, DbState, PhraseItem};
use crate::error::AppError;
use crate::srs::{apply_rating_phrase, Rating};
use crate::vocab::{self, AddPhraseInput};
use tauri::AppHandle;

#[tauri::command]
pub async fn add_phrase(app: AppHandle, input: AddPhraseInput) -> Result<PhraseItem, AppError> {
    let cfg = config::load_config()?;
    crate::commands::spawn_db(app, move |state| {
        Ok(vocab::add_or_merge_phrase(state, &cfg, input)?)
    })
    .await
}

#[tauri::command]
pub fn list_phrases(
    state: tauri::State<'_, DbState>,
    status: Option<String>,
) -> Result<Vec<PhraseItem>, AppError> {
    let conn = state.lock_read()?;
    Ok(db::list_phrases(&conn, status.as_deref())?)
}

#[tauri::command]
pub fn due_phrases(state: tauri::State<'_, DbState>) -> Result<Vec<PhraseItem>, AppError> {
    let conn = state.lock_read()?;
    Ok(db::due_phrases(&conn)?)
}

#[tauri::command]
pub fn review_phrase(
    state: tauri::State<'_, DbState>,
    id: String,
    rating: String,
) -> Result<PhraseItem, AppError> {
    let r = Rating::from_str(&rating)?;
    let conn = state.lock_write()?;
    let mut item = db::get_phrase(&conn, &id)?.ok_or_else(|| "phrase not found".to_string())?;
    apply_rating_phrase(&mut item, r);
    db::update_phrase_review(&conn, &item)?;
    Ok(item)
}

#[tauri::command]
pub fn set_phrase_status(
    state: tauri::State<'_, DbState>,
    id: String,
    status: String,
) -> Result<(), AppError> {
    let conn = state.lock_write()?;
    Ok(db::set_phrase_status(&conn, &id, &status)?)
}

#[tauri::command]
pub fn delete_phrase(state: tauri::State<'_, DbState>, id: String) -> Result<(), AppError> {
    let conn = state.lock_write()?;
    Ok(db::delete_phrase(&conn, &id)?)
}
